src/osd/PrimaryLogPG.cc

```diff
@@ -5989,6 +5990,61 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 
       break;
 
+    case CEPH_OSD_OP_SET_CHUNK:
+      ++ctx->num_write;
+      {
+	if (pool.info.is_tier()) {
+	  result = -EINVAL;
+	  break;
+	}
+	if (!obs.exists) {
+	  result = -ENOENT;
+	  break;
+	}
+	if (get_osdmap()->require_osd_release < CEPH_RELEASE_LUMINOUS) {
+	  result = -EOPNOTSUPP;
+	  break;
+	}
+
+	object_locator_t target_oloc;
+	uint64_t chunk_length;
+	try {
+	  ::decode(target_oloc, bp);
+	  ::decode(chunk_length, bp);
+	}
+	catch (buffer::error& e) {
+	  result = -EINVAL;
+	  goto fail;
+	}
+	
+	if (!chunk_length) {
+	  result = -EINVAL;
+	  goto fail;
+	}
+
+	for (uint64_t cursor = 0; cursor <= oi.size; cursor += chunk_length) {
+	  pg_t raw_pg;
+	  string chunk_name = soid.oid.name;
+	  chunk_name = chunk_name + "_" + to_string(cursor);
+	  get_osdmap()->object_locator_to_pg(chunk_name, target_oloc, raw_pg);
+	  hobject_t target(chunk_name, target_oloc.key, snapid_t(),
+			   raw_pg.ps(), raw_pg.pool(),
+			   target_oloc.nspace);
+	  chunk_info_t chunk_info;
+	  chunk_info.length = chunk_length;
+	  chunk_info.oid = target;
+	  chunk_info.flags = chunk_info_t::CHUNK_DIRTY;
+	  oi.manifest.chunk_map[cursor] = chunk_info;
+	}
+
+	oi.set_flag(object_info_t::FLAG_MANIFEST);
```

Comment:
I wonder if we should have something like
<pre>
if (!oi.manifest.is_chunked()) {
  oi.manifest.clear();
}
</pre>
so that the redirect target is cleared if it was a redirect (or any other future stuff that goes into the manifest is cleaned up).